x/stake/keeper.go

```diff
@@ -33,319 +33,315 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
 	// add the actual validator power sorted store
-	maxVal := k.GetParams(ctx).MaxValidators
-	iterator := store.ReverseIterator(subspace(ValidatorsKey)) // largest to smallest
-	for i := 0; ; i++ {
-		if !iterator.Valid() || i > int(maxVal-1) {
+	maxValidators := k.GetParams(ctx).MaxValidators
+	iterator = store.ReverseSubspaceIterator(ValidatorsByPowerKey) // largest to smallest
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() || i > int(maxValidators-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
-		var val Validator
-		err := k.cdc.UnmarshalBinary(bz, &val)
-		if err != nil {
-			panic(err)
-		}
-		if bytes.Equal(val.Address, address) {
-			return true
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+
+		// remove from ToKickOut group
+		toKickOut[string(validator.Address)] = nil
+
+		// also add to the current validators group
+		store.Set(GetValidatorsBondedKey(validator.PubKey), bz)
+
+		// MOST IMPORTANTLY, send back information that the current validator should be bonded
+		if bytes.Equal(updatedValidatorAddr, validator.Address) {
+			updatedIsBonded = true
 		}
+
 		iterator.Next()
 	}
 
-	return false
+	for _, value := range toKickOut {
+		if value == nil {
+			continue
+		}
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(value, &validator)
+		k.unbondValidator(ctx, store, validator)
+	}
+	return
 }
 
-// Is the address provided a part of the most recently saved validator group?
-func (k Keeper) IsRecentValidator(ctx sdk.Context, address sdk.Address) bool {
-	store := ctx.KVStore(k.storeKey)
-	if store.Get(GetRecentValidatorKey(address)) == nil {
-		return false
-	}
-	return true
+// perform all the store operations for when a validator status becomes unbonded
+func (k Keeper) unbondValidator(ctx sdk.Context, store sdk.KVStore, validator Validator) {
+	pool := k.GetPool(ctx)
+
+	// set the status
+	validator.Status = sdk.Unbonded
+	validator, pool = validator.UpdateSharesLocation(pool)
+	k.setPool(ctx, pool)
+
+	// save the now unbonded validator record
+	bz := k.cdc.MustMarshalBinary(validator)
+	store.Set(GetValidatorKey(validator.Address), bz)
+
+	// add to accumulated changes for tendermint
+	bz = k.cdc.MustMarshalBinary(validator.abciValidatorZero(k.cdc))
+	store.Set(GetTendermintUpdatesKey(validator.Address), bz)
+
+	// also remove from the Bonded Validators Store
+	store.Delete(GetValidatorsBondedKey(validator.PubKey))
+}
+
+// perform all the store operations for when a validator status becomes bonded
+func (k Keeper) bondValidator(ctx sdk.Context, store sdk.KVStore, validator Validator, pool Pool) {
+
+	// set the status
+	validator.Status = sdk.Bonded
+	validator, pool = validator.UpdateSharesLocation(pool)
+	k.setPool(ctx, pool)
+
+	// save the now bonded validator record to the three referened stores
+	bzVal := k.cdc.MustMarshalBinary(validator)
+	store.Delete(GetValidatorsBondedByPowerKey(validator, pool))
```

Comment:
Shouldn't this be overwritten anyways by the `store.Set` call two lines below?

Reply:
that's not always the case because the key is dependant on the power of the validator/among other things which may be getting updated within this function... so the new key in that store is not necessarily the same as the original key, but there should only be one... I just moved to earlier in the function before the validator is updated. (I think may have actually been a bug earlier)